### PR TITLE
Task 47 userlist

### PR DESCRIPTION
**Pull request by JulesMary**

### CODE REVIEW

File: app/src/components/chat/Chatrooms.js [hunk 1]
```diff
@@ -20,17 +21,31 @@ class Chatrooms extends React.Component {
 
     onSubmit = (value) => {
         const roomName = value;
-
-        if (roomName) {
+        const user = {
+            name: this.props.user.firstname + ' ' + this.props.user.lastname,
+            email: this.props.user.email,
+            role: 'ADMIN',
```
Comment (nilsjung) [inline app/src/components/chat/Chatrooms.js]: the roles should be defined global in `config.js` etc. like
```js
export const role = {
  admin: 'ADMIN',
   guest:  'GUEST',
...
}
```

File: app/src/components/login/Login.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ import LoginForm from './LoginForm';
 
 class Login extends React.Component {
     componentDidMount() {
+        console.log('logging out');
```
Comment (nilsjung) [inline app/src/components/login/Login.js]: `console.log`

File: app/src/components/chat/Chatrooms.js [hunk 3]
```diff
@@ -20,17 +21,31 @@ class Chatrooms extends React.Component {
...
+        };
+        console.log(user);
```
Comment (nilsjung) [inline app/src/components/chat/Chatrooms.js]: `console.log`

File: app/src/components/userList/list.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    constructor(props) {
+        super(props);
+        this.props.loadUsers(this.props.accessToken);
+        this.state = {
+            query: '',
+        };
+    }
+
+    handleClick = (user) => {
+        console.log(user);
```
Comment (nilsjung) [inline app/src/components/userList/list.js]: `console.log`

File: app/src/components/userList/list.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        const currentUser = {
+            name: this.props.user.firstname + ' ' + this.props.user.lastname,
+            email: this.props.user.email,
+            role: 'ADMIN',
```
Comment (nilsjung) [inline app/src/components/userList/list.js]: add global role conf

File: app/src/components/userList/list.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        this.props.handleItemClick({
+            chatroom: currentUser.name + ' + ' + chatPartner.name,
+            users: [currentUser, chatPartner],
+            token: this.props.accessToken,
+        });
+    };
+
+    handleChange = (event) => {
+        this.setState({ query: event.target.value });
+        const query = event.target.value.toLowerCase();
```
Comment (nilsjung) [inline app/src/components/userList/list.js]: take the query first, so you just have to access the event target one time:
```js
handleChange = (event) => {
  const query = event.target.value;
  this.setState({query})

  const users = this.props.users.map( (user) => {
    if (user.email.indexOf(query.toLowerCase()) !== -1) {
      return user.email
    }
  }
}

File: app/src/components/userList/list.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                        type="button"
+                        data-toggle="dropdown"
+                        aria-haspopup="true"
+                        aria-expanded="false"
+                    >
+                        Users
+                    </button>
+                    <div className="dropdown-menu" aria-labelledby={dropdownId}>
+                        <h6 className="dropdown-header">Search</h6>
+                        <input
```
Comment (nilsjung) [inline app/src/components/userList/list.js]: we have an input component defined

File: app/src/reducers/index.js [hunk 8]
```diff
@@ -84,6 +96,19 @@ export default function(state = initialState, action) {
         case LOAD_HISTORY:
             return { ...state, messages: action.chats };
 
+        case USERS:
+            const users = action.users.filter(
```
Comment (nilsjung) [inline app/src/reducers/index.js]: the content should be moved in a user-reducer so the switch-case just deligates to the reducer and has no logical stuff

File: app/src/reducers/userReducer.js [hunk 9]
```diff
@@ -6,9 +6,14 @@ export function setUserIdReducer(state, action) {
 
 export function isLoginSuccessfullReducer(state, action) {
     const { user, accessToken, chatrooms } = action;
+    const copiedState = Object.create(state);
```
Comment (nilsjung) [inline app/src/reducers/userReducer.js]: *just as a hint*: you can also use the spread-operator: `const copiedstate = {...state};`

Comment (nilsjung) [inline app/src/reducers/userReducer.js]: why is this necessary?

File: backend/model/chatroom.js [hunk 10]
```diff
@@ -22,6 +22,13 @@ var chatroomSchema = new mongoose.Schema({
             timestamp: Date,
         },
     ],
+    users: [
```
Comment (nilsjung) [inline backend/model/chatroom.js]: i have no idea how to model a mongoose schema, but i would add here the user schema, like
```js
const UserSchema = require('./User').schema;
...
const chatroomSchema = new mongoose.Schema({
  name: ...,
  ...
  users: [UserSchema],
});
```

File: backend/socket/socket.js [hunk 11]
```diff
@@ -9,41 +9,40 @@ const socket = (server) => {
...
 
     return io;
 };
 
-const messageService = async (data, connections) => {
-    const user = await token.verify(data.token);
+const messageService = (sock, data, connections) => {
     const chatroom = data.chatroom;
-
+    console.log(connections.length);
```
Comment (nilsjung) [inline backend/socket/socket.js]: `console.log`

File: backend/socket/socket.js [hunk 12]
```diff
@@ -9,41 +9,40 @@ const socket = (server) => {
...
+
+        sock.on('disconnect', () => disconnectService(connections));
     });
 
     return io;
 };
 
-const messageService = async (data, connections) => {
-    const user = await token.verify(data.token);
+const messageService = (sock, data, connections) => {
```
Comment (nilsjung) [inline backend/socket/socket.js]: if i am right, the argument `connections` is not used anymore?

File: backend/socket/socket.js [hunk 13]
```diff
@@ -9,41 +9,40 @@ const socket = (server) => {
...
+    sock.broadcast.emit('message', {
+        message: data.message,
+        user: data.user,
     });
 };
 
 const disconnectService = (connections) => {
     const index = connections.indexOf(socket);
     connections.splice(index, 1);
+    console.log(connections.length);
```
Comment (nilsjung) [inline backend/socket/socket.js]: `console.log`

File: app/src/reducers/userReducer.js (hunk 9 above)
Comment (JulesMary) [reply app/src/reducers/userReducer.js]: otherwise we would modify the state directly, since javascript holds the reference instead of a copy.. i think :D

File: backend/socket/socket.js (hunk 12 above)
Comment (JulesMary) [reply backend/socket/socket.js]: so damn right
